File: src/main/java/frc/robot/subsystems/swerve/module/ModuleIO.java
package frc.robot.subsystems.swerve.module;

import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Temperature;
import edu.wpi.first.units.measure.Voltage;

import static edu.wpi.first.units.Units.*;

import org.littletonrobotics.junction.AutoLog;

public interface ModuleIO {
    @AutoLog
    public static class ModuleIOInputs {
        public boolean driveConnected = false;
        public boolean driveOpenLoop = false;
        public Angle drivePosition = Radians.of(0);
        public AngularVelocity driveVelocity = RadiansPerSecond.of(0);
        public AngularVelocity driveGoal = RadiansPerSecond.of(0);
        public AngularVelocity driveSetpoint = RadiansPerSecond.of(0);
        public Voltage driveAppliedVolts = Volts.of(0);
        public Current driveCurrent = Amps.of(0);
        public Temperature driveTemp = Celsius.of(0);

        public boolean turnConnected = false;
        public boolean turnOpenLoop = false;
        public Angle turnPosition = Radians.of(0);
        public Angle turnGoal = Radians.of(0);
        public Angle turnSetpoint = Radians.of(0);
        public AngularVelocity turnVelocity = RadiansPerSecond.of(0);
        public Voltage turnAppliedVolts = Volts.of(0);
        public Current turnCurrent = Amps.of(0);
        public Temperature turnTemp = Celsius.of(0);

        public double[] odometryTimestamps = new double[] {};
        public double[] odometryDrivePositionsRad = new double[] {};
        public double[] odometryTurnPositionsRad = new double[] {};
    }

    /** Updates the set of loggable inputs. */
    public default void updateInputs(ModuleIOInputs inputs) {
    }

    /** Run the drive motor at the specified open loop value. */
    public default void setDriveOpenLoop(Voltage output) {
    }

    /** Run the turn motor at the specified open loop value. */
    public default void setTurnOpenLoop(Voltage output) {
    }

    /** Run the drive motor at the specified velocity. */
    public default void setDriveVelocity(AngularVelocity velocity) {
    }

    /** Run the turn motor to the specified rotation. */
    public default void setTurnPosition(Angle rotation) {
    }
}

File: src/main/kotlin/frc/robot/subsystems/swerve/module/ModuleIO.kt
package frc.robot.subsystems.swerve.module

import edu.wpi.first.units.Units.Amps
import edu.wpi.first.units.Units.Celsius
import edu.wpi.first.units.Units.Radians
import edu.wpi.first.units.Units.RadiansPerSecond
import edu.wpi.first.units.Units.Volts
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.AngularVelocity
import edu.wpi.first.units.measure.Current
import edu.wpi.first.units.measure.Temperature
import edu.wpi.first.units.measure.Voltage
import org.littletonrobotics.junction.AutoLog

interface ModuleIO {

    @AutoLog
    open class ModuleIOInputs {
        @JvmField var driveConnected: Boolean = false
        @JvmField var driveOpenLoop: Boolean = false
        @JvmField var drivePosition: Angle = Radians.of(0.0)
        @JvmField var driveVelocity: AngularVelocity = RadiansPerSecond.of(0.0)
        @JvmField var driveGoal: AngularVelocity = RadiansPerSecond.of(0.0)
        @JvmField var driveSetpoint: AngularVelocity = RadiansPerSecond.of(0.0)
        @JvmField var driveAppliedVolts: Voltage = Volts.of(0.0)
        @JvmField var driveCurrent: Current = Amps.of(0.0)
        @JvmField var driveTemp: Temperature = Celsius.of(0.0)

        @JvmField var turnConnected: Boolean = false
        @JvmField var turnOpenLoop: Boolean = false
        @JvmField var turnPosition: Angle = Radians.of(0.0)
        @JvmField var turnGoal: Angle = Radians.of(0.0)
        @JvmField var turnSetpoint: Angle = Radians.of(0.0)
        @JvmField var turnVelocity: AngularVelocity = RadiansPerSecond.of(0.0)
        @JvmField var turnAppliedVolts: Voltage = Volts.of(0.0)
        @JvmField var turnCurrent: Current = Amps.of(0.0)
        @JvmField var turnTemp: Temperature = Celsius.of(0.0)

        @JvmField var odometryTimestamps: DoubleArray = doubleArrayOf()
        @JvmField var odometryDrivePositionsRad: DoubleArray = doubleArrayOf()
        @JvmField var odometryTurnPositionsRad: DoubleArray = doubleArrayOf()
    }

    fun updateInputs(inputs: ModuleIOInputs) {}

    fun setDriveOpenLoop(output: Voltage) {}

    fun setTurnOpenLoop(output: Voltage) {}

    fun setDriveVelocity(velocity: AngularVelocity) {}

    fun setTurnPosition(rotation: Angle) {}
}
